src/index.js

```diff
@@ -110,18 +116,10 @@ export default class VueRouter {
 
     const history = this.history
 
-    if (history instanceof HTML5History) {
-      history.transitionTo(history.getCurrentLocation())
-    } else if (history instanceof HashHistory) {
-      const setupHashListener = () => {
-        history.setupListeners()
-      }
-      history.transitionTo(
-        history.getCurrentLocation(),
-        setupHashListener,
-        setupHashListener
-      )
+    const setupListeners = () => {
+      history.setupListeners()
     }
+    history.transitionTo(history.getCurrentLocation(), setupListeners, setupListeners)
```

Comment:
There's some new behavior here - previously `transitionTo()` would only get called during app initialization if using html5 or hash history, but now it is being called for abstract mode, too. I did so not because abstract mode needs this, but rather because it simplified the if/else code that existed before and appeared to have no harmful side effects.

Reply:
For this part, it's better to keep the previous behavior because abstract mode cannot always directly transition: e.g. during SSR whereas History and Hash can retrieve the current location on browsers

Reply:
I also realized we also need to keep the previous behavior for hash history for the hash event not to fire too early on some browsers

Reply:
>  it's better to keep the previous behavior because abstract mode cannot always directly transition:

👍 updated

> I also realized we also need to keep the previous behavior for hash history for the hash event not to fire too early on some browsers

That behavior still exists in this PR, as far as I can tell. HashHistory does not set up listeners until `setupListeners()` is called, and `setupListeners()` is not called until the first `history.transitionTo()` finishes up (as shown in the line of code above)

Reply:
The thing that's different here is that now that behavior for HashHistory is also applying to Html5History - beforehand the popstate listener was set up before the first transition, but it now is set up only after the first transition.

I think that new behavior is completely fine. My guess as to why HashHistory needed that special behavior is that hashchange events are fired as a result of programmatic navigation (via `location.hash = '#/new'`), whereas the browser **never** fires popstate events  never for programmatic navigation   (via `history.pushState(state, title, '/new')`)

Reply:
Oops I made the changes related to Abstract History weeks ago but forgot to push!! I commented and said "Updated" even though I didn't push. I have now pushed my changes.